### PR TITLE
Add missing test guards

### PR DESCRIPTION
**Pull request by makslevental**
- [x] iree_lit_test is missing filtering
- [x] runtime/src/iree/runtime/testdata isn't really testonly data, it is used by a sample and should be moved into that sample directly

### CODE REVIEW

File: runtime/src/iree/runtime/demo/CMakeLists.txt
Comment (makslevental) [inline runtime/src/iree/runtime/demo/CMakeLists.txt]: this isn't technically a test dir but down below there's a dep on a `iree::runtime::testdata::simple_mul_module_c`

Comment (benvanik) [reply runtime/src/iree/runtime/demo/CMakeLists.txt]: These aren't really tests and I'd not want to disable them based on that - they're binaries that ensure the system builds correctly. Maybe this should be moved to samples/, but since it's a strong indicator of a successful build (vs a unit/integration test) I'm not sure... I'll let @ScottTodd decide.

File: runtime/src/iree/vm/CMakeLists.txt
Comment (benvanik) [inline runtime/src/iree/vm/CMakeLists.txt]: We use bazel-to-cmake and need to make changes in the bazel files first - otherwise these will be overwritten the next time we generate them. You can run it with `python build_tools/bazel_to_cmake/bazel_to_cmake.py` and see what survives - some directories don't use it, but most do. Look for the giant block at the top of the file:
```
################################################################################
# Autogenerated by build_tools/bazel_to_cmake/bazel_to_cmake.py from           #
# runtime/src/iree/runtime/testdata/BUILD.bazel                                #
#                                                                              #
# Use iree_cmake_extra_content from iree/build_defs.oss.bzl to add arbitrary   #
# CMake-only content.                                                          #
#                                                                              #
# To disable autogeneration for this file entirely, delete this header.        #
################################################################################
```

There's ways of inserting the conditions (see some of the bazel files with extra content markers).

But, none of this is required - iree_cc_test disables itself unless IREE_BUILD_TESTS is set:
https://github.com/iree-org/iree/blob/63a2d141adf7b3911bc94c7f779e1896fb25aecb/build_tools/cmake/iree_cc_test.cmake#L54-L57

File: samples/static_library/CMakeLists.txt
Comment (benvanik) [inline samples/static_library/CMakeLists.txt]: iree_lit_test should disable itself ala iree_cc_test - iree_lit_test_suite does, but no clue why iree_lit_test doesn't (@ScottTodd was that intentional?) - I'd rather us make those rules consistent instead of doing this

**Review by benvanik — DISMISSED** [review]
I don't think most of this is needed given iree_cc_test behavior and we should make iree_lit_test behave the same, removing the need for those changes as well.

**Comment by makslevental** [comment]
@benvanik 

Much more below the surface than I expected. Thanks for the thorough pointers. This was an attempt to side-step a build issue in an alternative build path I was wanting to workout but now I'm thinking I'll just not side-step. 

But if there's some maintenance/chore work to be done here I'm happy to do that too. I.e., I'll let @ScottTodd weigh in and then you guys can tell me what should survive and I'll get it done.

File: samples/static_library/CMakeLists.txt
Comment (ScottTodd) [reply samples/static_library/CMakeLists.txt]: Yeah `iree_lit_test` is missing this code: https://github.com/iree-org/iree/blob/63a2d141adf7b3911bc94c7f779e1896fb25aecb/build_tools/cmake/iree_lit_test.cmake#L100-L103

Just an oversight - I only added a direct use of `iree_lit_test` recently. We had been exclusively using it through the wrapper `iree_lit_test_suite` before.

File: runtime/src/iree/runtime/demo/CMakeLists.txt
Comment (ScottTodd) [reply runtime/src/iree/runtime/demo/CMakeLists.txt]: Looks like the "testdata" folder there is leftover from style on previous projects where the line between "sample" and "test" was blurrier. In IREE now we have `IREE_BUILD_TESTS` and `IREE_BUILD_SAMPLES` that are separable.

Could rename `iree::runtime::testdata` to `iree::runtime::demodata` by renaming the directory or (my preference) just collapse that 'testdata' directory into the `demo` folder.

```
iree
  runtime
    demo
      hello_world_terse.c
      simple_mul.mlir
      CMakeLists.txt
```

then I'd also update the path in this comment (didn't find other uses but should confirm): https://github.com/iree-org/iree/blob/63a2d141adf7b3911bc94c7f779e1896fb25aecb/tools/iree-dump-instruments-main.c#L268

**Comment by ScottTodd** [comment]
Agreed that we should let the rules themselves do the filtering when possible. That also works with bazel_to_cmake so a bazel `cc_test` turns into a CMake `iree_cc_test` that automatically filters with no need for explicit `if(NOT IREE_BUILD_TESTS)` checks. There is some similar filtering applied for non-test targets (like test data or test library helpers) with the `testonly` tag: https://bazel.build/reference/be/common-definitions#:~:text=incompatible%20target%20skipping.-,testonly,-Boolean%3B%20nonconfigurable

Looks like this did uncover two real issues to fix
* `iree_test_test` is missing filtering
* `runtime/src/iree/runtime/testdata` isn't really testonly data, it is used by a sample and should be moved into that sample directly

**Comment by benvanik** [comment]
cool! sounds like some nice fixes - would be great to have a CI build config with samples/tests disabled to help catch some of these issues too!

**Comment by makslevental** [comment]
Okay I reverted my first n00b commit and added the early return in `iree_lit_test`. I'll get to moving `testdata` soon.

**Comment by benvanik** [comment]
neato - sorry that you peered into the darkness and were met with a colorful rainbow ;P
image

**Comment by makslevental** [comment]
@ScottTodd @benvanik I believe I have done the right things but I have not verified (because bazel...). Can you kick the tests again (or give me contrib access...).

**Comment by ScottTodd** [comment]
> @ScottTodd @benvanik I believe I have done the right things but I have not verified (because bazel...). Can you kick the tests again (or give me contrib access...).

Approved workflows and sent an invite to iree-org and the iree-triage team.

**Comment by makslevental** [comment]
> Approved workflows and sent an invite to iree-org and the iree-triage team.

Cool thanks. I'll figure out the fails tomorrow.

**Comment by ScottTodd** [comment]
> > Approved workflows and sent an invite to iree-org and the iree-triage team.
> 
> Cool thanks. I'll figure out the fails tomorrow.

Looks like the `//runtime/src/iree/runtime/testdata:simple_mul_module_c` target paths should be updated to `:simple_mul_module_c` now that the targets live side by side.

**Comment by makslevental** [comment]
> Looks like the `//runtime/src/iree/runtime/testdata:simple_mul_module_c` target paths should be updated to `:simple_mul_module_c` now that the targets live side by side.

Thanks for the tip - it probably would've taken me an hour to figure that out lol

File: runtime/src/iree/runtime/demo/BUILD.bazel
Comment (ScottTodd) [inline runtime/src/iree/runtime/demo/BUILD.bazel]: This will need to match the source code

```suggestion
    c_identifier = "iree_runtime_demo_simple_mul_module",
```

See this part of your diff in `runtime/src/iree/runtime/demo/hello_world_explained.c`
```diff
-  const iree_file_toc_t* module_file =
-      iree_runtime_testdata_simple_mul_module_create();
+  const iree_file_toc_t* module_file = iree_runtime_simple_mul_module_create();
```

(change further to `iree_runtime_demo_simple_mul_module_create`)

Comment (ScottTodd) [reply runtime/src/iree/runtime/demo/BUILD.bazel]: Those changes will help with the CI failures on https://github.com/iree-org/iree/actions/runs/9353849536/job/25745381896?pr=17540#step:4:113

(I'm now wondering why `build_all` succeeded... is that not building this demo?)

Comment (makslevental) [reply runtime/src/iree/runtime/demo/BUILD.bazel]: ah i thought i caught those (even though they're in BUILD)

Comment (ScottTodd) [reply runtime/src/iree/runtime/demo/BUILD.bazel]: I'd prefer to include `demo` in the name, so `iree_runtime_demo_simple_mul_module` instead of `iree_runtime_simple_mul_module` (that makes it seem like it's part of the public API and not something specific to a demo). Not too big a deal though.

**Comment by makslevental** [comment]
@ScottTodd I think I got it now?

**Review by ScottTodd — APPROVED** [review]

**Comment by ScottTodd** [comment]
Yep, looks good now! I think we can dismiss Ben's "requested changes" review and merge now.

**Comment by ScottTodd** [comment]
> cool! sounds like some nice fixes - would be great to have a CI build config with samples/tests disabled to help catch some of these issues too!

BTW I put some thoughts about a CI build on https://github.com/iree-org/iree/issues/17136#issuecomment-2145579825

**Pull request merged**
